servers/physics/space_sw.cpp

```diff
@@ -37,7 +37,7 @@ _FORCE_INLINE_ static bool _match_object_type_query(CollisionObjectSW *p_object,
 	if (p_object->get_type()==CollisionObjectSW::TYPE_AREA)
 		return p_type_mask&PhysicsDirectSpaceState::TYPE_MASK_AREA;
 
-	if ((p_object->get_layer_mask()&p_layer_mask)==0)
+	if (((p_object->get_layer_mask()&p_layer_mask)==0) && ((p_object->get_collision_mask()&p_layer_mask)==0))
```

Comment:
Objects on the same layer do not need to collide anyway. E.g.
* `a` - layer mask - `001`, collision mask - `010`
* `b` - layer mask - `001`, collision mask - `010`
* -> no collision :tada: (`a.layer & b.mask = 000`, `b.layer & a.mask = 000`)